# Phase 1 Project Data Cleaning / EDA

## Business Problem

This is the problem our project is specifying:

`"Your company is expanding in to new industries to diversify its portfolio. Specifically, they are interested in purchasing and operating airplanes for commercial and private enterprises, but do not know anything about the potential risks of aircraft. You are charged with determining which aircraft are the lowest risk for the company to start this new business endeavor. You must then translate your findings into actionable insights that the head of the new aviation division can use to help decide which aircraft to purchase."`

In this EDA, I do the following:
- Clean the Aviation_Data.csv dataset by:
    - Removing irrelevant measures
    - Removing duplicated records
    - Carefully removing or replacing null values
- Run a few different analyses to determine the lowest risk aircraft

## Importing the Data

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
plt.style.use('bmh')

We had an import issue that Nick solved by specifying the data type for columns 6, 7, and 28.

In [2]:
with open('../data/Aviation_Data.csv') as f:
    dtypes = {'Column6Name': 'str', 'Column7Name': 'str', 'Column28Name': 'str'}
    df = pd.read_csv(f, dtype=dtypes, low_memory=False)

In [3]:
df.head(2)

Event.Id Investigation.Type Accident.Number  Event.Date  \
0  20001218X45444           Accident      SEA87LA080  1948-10-24   
1  20001218X45447           Accident      LAX94LA336  1962-07-19   

          Location        Country Latitude Longitude Airport.Code  \
0  MOOSE CREEK, ID  United States      NaN       NaN          NaN   
1   BRIDGEPORT, CA  United States      NaN       NaN          NaN   

  Airport.Name  ... Purpose.of.flight Air.carrier Total.Fatal.Injuries  \
0          NaN  ...          Personal         NaN                  2.0   
1          NaN  ...          Personal         NaN                  4.0   

  Total.Serious.Injuries Total.Minor.Injuries Total.Uninjured  \
0                    0.0                  0.0             0.0   
1                    0.0                  0.0             0.0   

  Weather.Condition  Broad.phase.of.flight   Report.Status Publication.Date  
0               UNK                 Cruise  Probable Cause              NaN  
1               UNK                Unknown  Probable Cause       19-09-1996  

[2 rows x 31 columns]

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 90348 entries, 0 to 90347
Data columns (total 31 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Event.Id                88889 non-null  object 
 1   Investigation.Type      90348 non-null  object 
 2   Accident.Number         88889 non-null  object 
 3   Event.Date              88889 non-null  object 
 4   Location                88837 non-null  object 
 5   Country                 88663 non-null  object 
 6   Latitude                34382 non-null  object 
 7   Longitude               34373 non-null  object 
 8   Airport.Code            50249 non-null  object 
 9   Airport.Name            52790 non-null  object 
 10  Injury.Severity         87889 non-null  object 
 11  Aircraft.damage         85695 non-null  object 
 12  Aircraft.Category       32287 non-null  object 
 13  Registration.Number     87572 non-null  object 
 14  Make                    88826 non-null

Okay, so we've got 90,348 records and 31 columns. Let's clean this up a bit.

## Cleaning the Data

#### Keeping / Removing Columns for These Reasons

Keeping
- Event ID and Accident Number: keeping for now to screen for duplicates in next step*
- Event.Date: might be useful
- Location and Country: might be useful
- Injury Severity, Aircraft Damage: will be useful
- Aircraft Category: Shows many records not involving airplanes, which we can remove later*
- Make and Model: will be useful
- Amateur Built: can probably use to remove from data. About 8.5k out of 90 are listed as amateur built*
- Number of Engines: will need further investigation. Some show 0 engines, most show one. I would imagine we would only be looking at 2+ engine planes for enterprise use*
- Engine Type: useful
- FAR Description: Stands for Federal Aviation Regulation description. Might be useful, so keep for now.*
- Purpose of Flight: useful for determining cause of accident
- Total Injuries columns: useful
- Weather Conditions: will help determine possible cause
- Report Status: May be helpful for us

*Return to these measures for further cleaning

Removing
- Investigation Type: upon doing a .values_count(), we see it is not useful
- Latitude and Longitude: mostly null, also redundant since we have Location (city, state)
- Airport.Code and Airport Name: about half null, also irrelevant to determining safety
- Registration Number: irrelevant
- Schedule: Mostly null, also irrelevant
- Air Carrier: mostly null, aldo irrelevant
- Broad Phase of Flight: Enough null values for a categorical data point that we should exclude
- Publication Date: irrelevant when data was published


In [5]:
df.drop(['Investigation.Type','Latitude', 'Longitude', 'Airport.Code', 
         'Airport.Name', 'Registration.Number', 'Schedule', 'Air.carrier', 
         'Broad.phase.of.flight', 'Publication.Date'], axis=1, inplace=True)

In [6]:
df.head(2)

Event.Id Accident.Number  Event.Date         Location        Country  \
0  20001218X45444      SEA87LA080  1948-10-24  MOOSE CREEK, ID  United States   
1  20001218X45447      LAX94LA336  1962-07-19   BRIDGEPORT, CA  United States   

  Injury.Severity Aircraft.damage Aircraft.Category     Make     Model  ...  \
0        Fatal(2)       Destroyed               NaN  Stinson     108-3  ...   
1        Fatal(4)       Destroyed               NaN    Piper  PA24-180  ...   

  Number.of.Engines    Engine.Type FAR.Description Purpose.of.flight  \
0               1.0  Reciprocating             NaN          Personal   
1               1.0  Reciprocating             NaN          Personal   

  Total.Fatal.Injuries  Total.Serious.Injuries  Total.Minor.Injuries  \
0                  2.0                     0.0                   0.0   
1                  4.0                     0.0                   0.0   

   Total.Uninjured  Weather.Condition   Report.Status  
0              0.0                UNK  Probable Cause  
1              0.0                UNK  Probable Cause  

[2 rows x 21 columns]

In [7]:
df[df['Event.Id'].isna()].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1459 entries, 64030 to 90097
Data columns (total 21 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Event.Id                0 non-null      object 
 1   Accident.Number         0 non-null      object 
 2   Event.Date              0 non-null      object 
 3   Location                0 non-null      object 
 4   Country                 0 non-null      object 
 5   Injury.Severity         0 non-null      object 
 6   Aircraft.damage         0 non-null      object 
 7   Aircraft.Category       0 non-null      object 
 8   Make                    0 non-null      object 
 9   Model                   0 non-null      object 
 10  Amateur.Built           0 non-null      object 
 11  Number.of.Engines       0 non-null      float64
 12  Engine.Type             0 non-null      object 
 13  FAR.Description         0 non-null      object 
 14  Purpose.of.flight       0 non-null 

Getting all null values in rows where Event ID is null. Let's drop those records

In [8]:
df.dropna(subset=['Event.Id'], inplace=True)

Time to look for duplicates

In [9]:
# New df containing duplicated Event ID's

df_duplicates1 = df[df.duplicated(subset=['Event.Id'], keep=False) == True].sort_values('Event.Id')

df_duplicates1

Event.Id Accident.Number  Event.Date       Location  \
45562  20001204X00086     LAX99LA086B  1999-01-29     MADERA, CA   
45564  20001204X00086     LAX99LA086A  1999-01-29     MADERA, CA   
45703  20001205X00276     CHI99IA100B  1999-03-02     SALINA, KS   
45704  20001205X00276     CHI99IA100A  1999-03-02     SALINA, KS   
45716  20001205X00305     DEN99LA047B  1999-03-05     DENVER, CO   
...               ...             ...         ...            ...   
90255  20221121106336      WPR23LA041  2022-11-18  Las Vegas, NV   
90256  20221122106340      DCA23WA071  2022-11-18    Marrakech,    
90257  20221122106340      DCA23WA071  2022-11-18    Marrakech,    
90272  20221123106354      WPR23LA045  2022-11-22  San Diego, CA   
90273  20221123106354      WPR23LA045  2022-11-22  San Diego, CA   

             Country Injury.Severity Aircraft.damage Aircraft.Category  \
45562  United States       Non-Fatal           Minor               NaN   
45564  United States       Non-Fatal     Substantial               NaN   
45703  United States        Incident             NaN               NaN   
45704  United States        Incident             NaN               NaN   
45716  United States       Non-Fatal           Minor               NaN   
...              ...             ...             ...               ...   
90255  United States       Non-Fatal     Substantial          Airplane   
90256        Morocco             NaN             NaN               NaN   
90257        Morocco             NaN             NaN               NaN   
90272  United States       Non-Fatal     Substantial        Helicopter   
90273  United States       Non-Fatal     Substantial        Helicopter   

                    Make            Model  ... Number.of.Engines  \
45562             Coelho              RV4  ...               1.0   
45564             Cessna             150L  ...               1.0   
45703           Lockheed  L-1O11-385-1-15  ...               3.0   
45704  Mcdonnell Douglas            DC-10  ...               3.0   
45716         Swearingen          SA226TC  ...               2.0   
...                  ...              ...  ...               ...   
90255             CESSNA             172M  ...               1.0   
90256             AIRBUS             A320  ...               NaN   
90257             CESSNA              525  ...               NaN   
90272           SIKORSKY           UH-60A  ...               2.0   
90273           SIKORSKY           MH-60R  ...               2.0   

         Engine.Type FAR.Description          Purpose.of.flight  \
45562  Reciprocating             NaN                   Personal   
45564  Reciprocating             NaN                   Personal   
45703      Turbo Jet             NaN                    Unknown   
45704      Turbo Jet             NaN                    Unknown   
45716     Turbo Prop             NaN                    Unknown   
...              ...             ...                        ...   
90255            NaN             091              Instructional   
90256            NaN            NUSC                        NaN   
90257            NaN            NUSC                        NaN   
90272            NaN             091              Instructional   
90273            NaN            ARMF  Public Aircraft - Federal   

      Total.Fatal.Injuries  Total.Serious.Injuries  Total.Minor.Injuries  \
45562                  0.0                     0.0                   0.0   
45564                  0.0                     0.0                   0.0   
45703                  0.0                     0.0                   0.0   
45704                  0.0                     0.0                   0.0   
45716                  0.0                     0.0                   0.0   
...                    ...                     ...                   ...   
90255                  0.0                     0.0                   0.0   
90256                  0.0                     0.0                   0.0   
9

Looks like we found some misentries. The makes and models of these aircraft are disagreeing for seemingly the same accidents, which has created these duplicate records. Unfortunately, it's impossible for us to tell which record is accurate. To avoid bad data messing with our results, we will need to exclude ALL of these records.

In [10]:
df.drop_duplicates(subset=['Event.Id'], keep=False, inplace=True)

In [11]:
# Checking our work

df.duplicated(subset=['Event.Id']).value_counts()

False    87015
dtype: int64

In [12]:
# Let's check along the Accident Number column

df.duplicated(subset=['Accident.Number']).value_counts()

False    87015
dtype: int64

Looks good. For cleanliness, let's drop the Accident Number column. We can keep the Event ID for indexing and to allow others to check for duplicates.

In [13]:
df.drop(['Accident.Number'], axis=1, inplace=True)

Now let's look at Aircraft Category and exclude any non-airplane records

In [14]:
# These are the records we need to drop

df[(df['Aircraft.Category'] != 'Airplane') & (df['Aircraft.Category'].isna() == False)]

Event.Id  Event.Date           Location        Country  \
16     20020917X01962  1982-01-02         MIDWAY, UT  United States   
19     20020917X02339  1982-01-02          MIAMI, FL  United States   
22     20020917X01657  1982-01-02       CHAMBLEE, GA  United States   
46     20020917X02157  1982-01-06  MAMMOTH LAKES, CA  United States   
62     20020917X02247  1982-01-09      CALISTOGA, CA  United States   
...               ...         ...                ...            ...   
90301  20221204106407  2022-12-04        Beverly, MA  United States   
90313  20221209106435  2022-12-07         Waimea, HI  United States   
90315  20221212106442  2022-12-08         La Sal, UT  United States   
90322  20221213106449  2022-12-11          Kildare,         Ireland   
90326  20221213106456  2022-12-12        Redding, CA  United States   

      Injury.Severity Aircraft.damage Aircraft.Category  \
16          Non-Fatal       Destroyed        Helicopter   
19          Non-Fatal     Substantial        Helicopter   
22          Non-Fatal     Substantial        Helicopter   
46          Non-Fatal     Substantial        Helicopter   
62          Non-Fatal     Substantial            Glider   
...               ...             ...               ...   
90301           Fatal     Substantial         Gyrocraft   
90313           Minor     Substantial        Helicopter   
90315       Non-Fatal     Substantial        Helicopter   
90322             NaN       Destroyed        Helicopter   
90326           Minor     Substantial        Helicopter   

                              Make               Model Amateur.Built  \
16                         Enstrom                280C            No   
19                           Smith  WCS-222 (BELL 47G)            No   
22                            Bell              206L-1            No   
46                    Aerospatiale             SA-316B            No   
62                      Schleicher              ASW 20            No   
...                            ...                 ...           ...   
90301            ROTORSPORT UK LTD             CAVALON            No   
90313                   EUROCOPTER           EC 130 B4            No   
90315                       HUGHES                369D            No   
90322  ROBINSON HELICOPTER COMPANY              R44 II            No   
90326           AIRBUS HELICOPTERS           EC 130 T2            No   

       Number.of.Engines    Engine.Type                FAR.Description  \
16                   1.0  Reciprocating      Part 91: General Aviation   
19                   1.0  Reciprocating      Part 91: General Aviation   
22                   1.0    Turbo Shaft  Part 135: Air Taxi & Commuter   
46                   1.0    Turbo Shaft      Part 91: General Aviation   
62                   0.0        Unknown      Part 91: General Aviation   
...                  ...            ...                            ...   
90301                1.0            NaN                            091   
90313                1.0            NaN                            135   
90315                1.0            NaN                            091   
90322                1.0            NaN                           NUSN   
90326                1.0            NaN                            135   

      Purpose.of.flight  Total.Fatal.Injuries  Total.Serious.Injuries  \
16             Personal                   0.0                     0.0   
19             Personal                   0.0                     0.0   
22              Unknown                   0.0                     0.0   
46             Business                   0.0                     0.0   
62             Personal                   0.0                     0.0   
...                 ...                   ...                     ...   
90301          Personal                   1.0                     0.0   
90313               NaN                   0.0                     1.0   
90315    Other Work Use                 

In [15]:
# Get their indexes and drop

nonplane_indexes = df[(df['Aircraft.Category'] != 'Airplane') 
                      & (df['Aircraft.Category'].isna() == False)].index

df.drop(index=nonplane_indexes, inplace=True)

In [16]:
# Now let's drop that column

df.drop(['Aircraft.Category'], axis=1, inplace=True)

Let's look further into Engine Counts. We can probably drop all records for single engine aircraft, as our company probably will not be using prop planes. But first - what's going on with the 0 engines?

In [17]:
df['Number.of.Engines'].value_counts()

1.0    65089
2.0    10494
0.0      700
3.0      427
4.0      388
6.0        1
8.0        1
Name: Number.of.Engines, dtype: int64

In [18]:
df[df['Number.of.Engines'] == 0]

Event.Id  Event.Date         Location        Country  \
1505   20020917X03932  1982-06-09   SHREWSBURY, PA  United States   
3606   20001214X42064  1983-01-02    INDIANOLA, IA  United States   
3659   20001214X42066  1983-01-08    GREENWOOD, MO  United States   
3951   20001214X42143  1983-02-21     WOODBINE, MD  United States   
4093   20001214X42553  1983-03-12  BUENA VISTA, PA  United States   
...               ...         ...              ...            ...   
63301  20070907X01321  2007-09-02   HUTCHINSON, KS  United States   
63538  20071029X01673  2007-10-08  ALBUQUERQUE, NM  United States   
63634  20071030X01689  2007-10-26       Salida, CO  United States   
64890  20080827X01334  2008-07-05       Beloit, WI  United States   
77176  20150616X50938  2015-06-14    Pell City, AL  United States   

      Injury.Severity Aircraft.damage                  Make  \
1505        Non-Fatal     Substantial             Scheicher   
3606        Non-Fatal             NaN         Balloon Works   
3659        Non-Fatal     Substantial         Balloon Works   
3951        Non-Fatal     Substantial   Scheibe Flugzeugbau   
4093        Non-Fatal     Substantial         Burkhart Grob   
...               ...             ...                   ...   
63301       Non-Fatal     Substantial              Slingsby   
63538        Fatal(1)     Substantial              Aerostar   
63634        Fatal(1)       Destroyed         Schempp-hirth   
64890       Non-Fatal     Substantial  AB Sportine Aviacija   
77176       Non-Fatal     Substantial             WEATHERLY   

                   Model Amateur.Built  Number.of.Engines    Engine.Type  \
1505                 K8B            No                0.0        Unknown   
3606          FIREFLY 7B            No                0.0        Unknown   
3659        FIRE FLY 7-B            No                0.0        Unknown   
3951          L SPATZ-55            No                0.0        Unknown   
4093       G10Z ASTIR CS            No                0.0        Unknown   
...                  ...           ...                ...            ...   
63301  Swallow Type T.45            No                0.0            NaN   
63538              S-66A            No                0.0            NaN   
63634      Ventus B/16.6            No                0.0            NaN   
64890          Genesis 2            No                0.0            NaN   
77176               201B            No                0.0  Reciprocating   

      FAR.Description   Purpose.of.flight  Total.Fatal.Injuries  \
1505              NaN            Personal                   0.0   
3606              NaN            Personal                   0.0   
3659              NaN       Instructional                   0.0   
3951              NaN            Personal                   0.0   
4093              NaN            Personal                   0.0   
...               ...                 ...                   ...   
63301             NaN          Glider Tow                   NaN   
63538             NaN            Personal                   1.0   
63634             NaN            Personal                   1.0   
64890             091            Personal                   0.0   
77176             137  Aerial Application                   0.0   

       Total.Serious.Injuries  Total.Minor.Injuries  Total.Uninjured  \
1505                      0.0                   0.0              1.0   
3606                      1.0                   0.0              1.0   
3659                      0.0                   0.0              2.0   
3951                      0.0                   0.0              1.0   
4093                      0.0                   0.0              1.0   
...                       ...                   ...              ...   
63301                     1.0                   NaN              NaN   
63538                     2.0                   NaN              2.0   
63634                     NaN                   NaN          

With a bit of googling, we find that these are gliders and balloons. Exclude!

In [19]:
# Get their indexes and drop

engines_1and0_indexes = df[(df['Number.of.Engines'] == 0) 
                      | (df['Number.of.Engines'] == 1)].index

df.drop(index=engines_1and0_indexes, inplace=True)

Now let's drop any amateur-built planes. We certainly are not in the market for those. 

In [20]:
# Get their indexes and drop

amateur_indexes = df[(df['Amateur.Built'] == 'Yes')].index

df.drop(index=amateur_indexes, inplace=True)

In [21]:
# Now let's drop that column

df.drop(['Amateur.Built'], axis=1, inplace=True)

Where Injury Severity is listed as 'Incident', Aircraft Damage is null or Minor and Injuries are null, zero, or low.

Where Injury Severity is null, Injuries are null, zero, or low. 

Let's drop these records.

In [22]:
df[(df['Injury.Severity'] == 'Incident') | (df['Injury.Severity'].isna())]

Event.Id  Event.Date        Location        Country  \
79     20020917X01897  1982-01-12     CHICAGO, IL  United States   
80     20020917X01765  1982-01-12  CLARKSBURG, WV  United States   
119    20020917X01766  1982-01-19  WASHINGTON, DC  United States   
131    20020917X02334  1982-01-20    SAN JOSE, CA  United States   
149    20020917X01767  1982-01-22  LOUISVILLE, KY  United States   
...               ...         ...             ...            ...   
90321  20221212106440  2022-12-11  Fox Island, NY  United States   
90333  20221215106462  2022-12-15   Patterson, LA  United States   
90338  20221219106472  2022-12-18     Kahului, HI  United States   
90344  20221227106494  2022-12-26     Hampton, NH  United States   
90346  20221227106498  2022-12-26      Morgan, UT  United States   

      Injury.Severity Aircraft.damage                 Make      Model  \
79           Incident             NaN             Lockheed     L-1011   
80           Incident           Minor              Embraer  EMB-110P1   
119          Incident           Minor         De Havilland  DHC-6-300   
131          Incident           Minor                Piper  PA-31-350   
149          Incident             NaN         Dassault/sud  FALCON 20   
...               ...             ...                  ...        ...   
90321             NaN             NaN  ROBINSON HELICOPTER   R22 BETA   
90333             NaN             NaN                 BELL     206-L4   
90338             NaN             NaN               AIRBUS   A330-243   
90344             NaN             NaN             BELLANCA       7ECA   
90346             NaN             NaN               CESSNA       210N   

       Number.of.Engines    Engine.Type                FAR.Description  \
79                   3.0      Turbo Fan          Part 121: Air Carrier   
80                   2.0     Turbo Prop  Part 135: Air Taxi & Commuter   
119                  2.0     Turbo Prop      Part 91: General Aviation   
131                  2.0  Reciprocating      Part 91: General Aviation   
149                  2.0      Turbo Fan          Part 121: Air Carrier   
...                  ...            ...                            ...   
90321                NaN            NaN                            091   
90333                NaN            NaN                            135   
90338                NaN            NaN                            121   
90344                NaN            NaN                            NaN   
90346                NaN            NaN                            091   

         Purpose.of.flight  Total.Fatal.Injuries  Total.Serious.Injuries  \
79                 Unknown                   0.0                     0.0   
80                 Unknown                   0.0                     0.0   
119                  Ferry                   0.0                     0.0   
131    Executive/corporate                   0.0                     0.0   
149                Unknown                   0.0                     0.0   
...                    ...                   ...                     ...   
90321        Instructional                   0.0                     0.0   
90333                  NaN                   0.0                     0.0   
90338                  NaN                   0.0                     0.0   
90344                  NaN                   0.0                     0.0   
90346             Personal                   0.0                     0.0   

       Total.Minor.Injuries  Total.Uninjured Weather.Condition   Report.Status  
79                      0.0            149.0               UNK  Probable Cause  
80                      0.0              2.0               VMC  Probable Cause  
119                     0.0              1.0               IMC  Probable Cause  
131                     0.0              2.0               VMC  Probable Cause  
149                     0.0              2.0               VMC  Probable Cause  
...                     ...    

In [23]:
# Get their indexes and drop

incident_null_indexes = df[(df['Injury.Severity'] == 'Incident') | (df['Injury.Severity'].isna())].index

df.drop(index=incident_null_indexes, inplace=True)

Let's drop records where Make or Model are null

In [24]:
null_make_model_index = df[(df['Make'].isna()) | (df['Model'].isna())].index

df.drop(index=null_make_model_index, inplace=True)

In [29]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 13670 entries, 4 to 90347
Data columns (total 18 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Event.Id                13670 non-null  object 
 1   Event.Date              13670 non-null  object 
 2   Location                13646 non-null  object 
 3   Country                 13606 non-null  object 
 4   Injury.Severity         13670 non-null  object 
 5   Aircraft.damage         12332 non-null  object 
 6   Make                    13670 non-null  object 
 7   Model                   13670 non-null  object 
 8   Number.of.Engines       11150 non-null  float64
 9   Engine.Type             10724 non-null  object 
 10  FAR.Description         5124 non-null   object 
 11  Purpose.of.flight       10417 non-null  object 
 12  Total.Fatal.Injuries    12072 non-null  float64
 13  Total.Serious.Injuries  11625 non-null  float64
 14  Total.Minor.Injuries    11487 non-null

Now with cleaner data set I can wrap my head around some potential questions. In general, I'd like to know which Makes, Models, Engine Counts, and Engine Types are most/least involved in accidents, which I would like to measure by Injury Severity, Aircraft Damage, and Fatal and Serious Injury count.

Before going further, I'd like to try to fill in some null values for Engine Count and Engine Type by making a dictionary of existing counts and types by Model. Let's see if it works.

In [25]:
# Starting number

df['Number.of.Engines'].count()

9633

In [26]:
# Creating a dictionary using Model as keys and Number of Engines (non-null) as values

engine_count_dict = df[(df['Number.of.Engines'].isna() == 
                        False)].set_index('Model')['Number.of.Engines'].to_dict()

In [27]:
# Using this dictionary to fill in some null values in Number of Engines

df['Number.of.Engines'] = df['Number.of.Engines'].fillna(df['Model'].map(engine_count_dict))

In [30]:
# Ending number. Got another 1500!

df['Number.of.Engines'].count()

11150

Now let's do this for Engine Type.

In [32]:
# Starting number

df['Engine.Type'].count()

10724

In [34]:
# {Model: Engine Type}, then use to fill in nulls. Got another 2000!

engine_type_dict = df[(df['Engine.Type'].isna() == 
                        False)].set_index('Model')['Engine.Type'].to_dict()

df['Engine.Type'] = df['Engine.Type'].fillna(df['Model'].map(engine_type_dict))

df['Engine.Type'].count()

12828

In [35]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 13670 entries, 4 to 90347
Data columns (total 18 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Event.Id                13670 non-null  object 
 1   Event.Date              13670 non-null  object 
 2   Location                13646 non-null  object 
 3   Country                 13606 non-null  object 
 4   Injury.Severity         13670 non-null  object 
 5   Aircraft.damage         12332 non-null  object 
 6   Make                    13670 non-null  object 
 7   Model                   13670 non-null  object 
 8   Number.of.Engines       11150 non-null  float64
 9   Engine.Type             12828 non-null  object 
 10  FAR.Description         5124 non-null   object 
 11  Purpose.of.flight       10417 non-null  object 
 12  Total.Fatal.Injuries    12072 non-null  float64
 13  Total.Serious.Injuries  11625 non-null  float64
 14  Total.Minor.Injuries    11487 non-null

In [38]:
df[(df['Injury.Severity'] != 'Non-Fatal') & (df['Total.Fatal.Injuries'].isna())]

Event.Id  Event.Date                            Location  \
41240  20010711X01373  1997-02-08              NURNBERG/GERF, Germany   
41339  20010711X01374  1997-03-07  MEDELLIN, Central African Republic   
41656  20001208X07697  1997-04-27                        JINAN, CHINA   
42287  20001208X08352  1997-07-21           SOUTHERN THAILA, Thailand   
42847  20001208X08992  1997-10-09                       HARBIN, CHINA   
43692  20010711X01369  1998-04-12                  ALMATY, Kazakhstan   
44014  20001211X10045  1998-05-29                       BOMBAY, INDIA   
44608  20001211X10826  1998-08-05           SEOUL, KOREA, REPUBLIC OF   
45560  20001204X00045  1999-01-28                       NAPLES, ITALY   
45711  20001205X00299  1999-03-04                    BIARRITZ, FRANCE   
45764  20001205X00297  1999-03-15          POHANG, KOREA, REPUBLIC OF   
45978  20001205X00483  1999-04-24                  Georgetown, Guyana   
46927  20010330X00670  1999-08-21                    ISTANBUL, Turkey   
47191  20001212X19734  1999-09-23                   BANGKOK, THAILAND   
48024  20001212X20608  2000-03-20                      AREQUIPA, PERU   
48384  20010105X00025  2000-05-17                             , Gabon   
49914  20010329X00665  2001-01-06                       Dundo, Angola   
50235  20020104X00030  2001-03-17                     Lubango, Angola   
50342  20010904X01864  2001-04-08                   Helsinki, Finland   
51807  20020104X00027  2001-11-20                      TAIPEI, TAIWAN   
51838  20011128X02308  2001-11-27              Port Harcourt, Nigeria   
52141  20020226X00261  2002-02-17                 SAN JUAN, ARGENTINA   
52277  20020322X00386  2002-03-18              BELO HORIZONTE, BRAZIL   
52992  20020719X01171  2002-07-06                       PARIS, FRANCE   
54159  20031008X01681  2003-02-11               IKARIA ISLAND, Greece   
54284  20040528X00699  2003-03-12                        JOHANNESBURG   
54287  20040528X00698  2003-03-12               AUCKLAND, NEW ZEALAND   
55975  20040202X00143  2003-12-13                          LIMA, PERU   
56007  20040202X00145  2003-12-19                   LIBREVILLE, GABON   
56081  20040202X00147  2004-01-03              Sharm El Sheikh, Egypt   
56253  20080421X00521  2004-02-20                      CANNES, FRANCE   
56991  20040817X01235  2004-07-06              AMSTERDAM, NETHERLANDS   
57112  20040722X01041  2004-07-21                 Mexico City, Mexico   
57232  20040812X01208  2004-08-07              VIRU-VIRU INT, BOLIVIA   
57530  20041018X01645  2004-09-19                     Kolari, Finland   
57798  20041209X01954  2004-11-07       SHARJAH, UNITED ARAB EMIRATES   
58184  20050926X01529  2005-02-11              Achmad Yani, Indonesia   
58264  20050915X01467  2005-03-06                      VERADERO, CUBA   
58474  20050617X00805  2005-04-16                TERENGGANU, MALAYSIA   
58506  20050628X00886  2005-04-23                  STEINFELD, GERMANY   
58751  20050817X01260  2005-05-31                  JAKARTA, INDONESIA   
59826  20051109X01810  2005-11-07                     TAITUNG, TAIWAN   
59965  20060208X00178  2005-12-15                       MACAE, BRAZIL   
60549  20070814X01180  2006-04-27                   San Paulo, Brazil   
60573  20060705X00864  2006-05-01                       Amman, Jordan   
60804  20060801X01067  2006-06-07                       PARIS, FRANCE   
61413  20060918X01362  2006-09-14              SUBIC BAY, PHILIPPINES   
61561  20061020X01543  2006-10-13                 Dolsenhain, Germany   
61714  20061122X01703  2006-11-18                   Leticia, Colombia   
62445  20070531X00673  2007-05-01        PETERBOROUGH, UNITED KINGDOM   
62585  20071108X01758  2007-05-25                ACARICUARA, COLOMBIA   
63178  20070905X01309  2007-08-15            WIENER NEUSTADT, AUSTRIA   
63398  20071001X01488  2007-09-16             BANBURY, UNITED KINGDOM   

                        Country Injury.Severity Aircraft.damage  \
412

In [39]:
df['Report.Status'].value_counts()

Probable Cause                                                                                                                                                                                                                                                                                                                                                             7714
Foreign                                                                                                                                                                                                                                                                                                                                                                    1354
<br /><br />                                                                                                                                                                                                                                                            